### PR TITLE
Feature: Yaml JdbcIO partitioning for ReadRows

### PR DESCRIPTION
**Pull request by shubham-j-sde**
addresses #34349 Beam Yaml JdbcIO partitioning

- [x] Extend [ReadRows](https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L730) to accept partitioning.
- [x] Adding test in [JdbcIOTest](https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java).
- [ ] Add Integration Test in [JdbcIOIT](https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java#L235)
- [ ] Extend [JdbcReadSchemaTransformProvider](https://github.com/Polber/beam/blob/master/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java) to accept this as a parameter.
- [ ] Update the mapping in [standard_io.yaml](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/yaml/standard_io.yaml#L195)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by shubham-j-sde** [comment]
@svetakvsundhar, I have extended said class per my understanding, please let me know for any required changes. proceeding to write tests.
Thank you

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 1]
```diff
@@ -889,6 +927,305 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+    public ReadRowsWithPartitions<PartitionColumnT> withTable(String tableName) {
+      checkNotNull(tableName, "table can not be null");
+      return toBuilder().setTable(tableName).build();
+    }
+
+    public ReadRowsWithPartitions<PartitionColumnT> withSchema(Schema schema) {
+      return toBuilder().setSchema(schema).build();
+    }
+
+    private static final int EQUAL = 0;
```
Comment (svetakvsundhar) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:1072]: Can we add a comment as to what this variable is for?

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 2]
```diff
@@ -515,6 +515,110 @@ public void testIncompatibleSchemaThrowsError() {
...
+            .withTable(READ_TABLE_NAME)
+            .withNumPartitions(0)
+            .withPartitionColumn("id")
+            .withLowerBound(0L)
+            .withUpperBound(1000L));
+    pipeline.run();
+  }
+
+  @Test
+  public void testReadRowsPartitionsLowerBoundIsMoreThanUpperBound() {
```
Comment (svetakvsundhar) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java:586]: Can we also add a test to test a null partition column and a null table name?

**Review by svetakvsundhar — CHANGES_REQUESTED** [review]
Thanks for working on this, had a couple of comments.

R: @Abacn for any more specifics on this Jdbc IO change

**Comment by svetakvsundhar** [comment]
We will also want to add an integration test [here](https://github.com/apache/beam/blob/3595a3303df6d2e4c8e77482ca8120994d0d5881/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java#L235), testing the newly added partitioning property.

**Comment by claudevdm** [comment]
Hi, is this change necessary? The current ReadWithPartitions already has the ability to output a row, by specifying withRowOutput [1]. Here is an example of it being used in a separate cross language transform [2]

[1] https://github.com/apache/beam/blob/3595a3303df6d2e4c8e77482ca8120994d0d5881/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L1432
[2] https://github.com/apache/beam/blob/3595a3303df6d2e4c8e77482ca8120994d0d5881/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcSchemaIOProvider.java#L131

**Comment by svetakvsundhar** [comment]
@claudevdm -- it seems necessary as the end goal of this, as described in https://github.com/apache/beam/issues/34349 is to expose this to Beam YAML, which uses the ReadRows Transform when reading from JDBC.

**Comment by shubham-j-sde** [comment]
Thank you for the inputs. on it!

**Comment by claudevdm** [comment]
> @claudevdm -- it seems necessary as the end goal of this, as described in #34349 is to expose this to Beam YAML, which uses the ReadRows Transform when reading from JDBC.

This PR seems like adds a new ReadRowsWithPartitions, which achieves the same as ReadWithPartitions.withRowOutput(). Can you help me understand why are we adding a new ReadRowsWithPartitions with almost the exact same code as ReadWithPartitions?

ReadWithPartitions.withRowOutput() can be used from  https://github.com/apache/beam/blob/3595a3303df6d2e4c8e77482ca8120994d0d5881/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java#L49 can it not?

**Comment by shubham-j-sde** [comment]
changes were moved on to PR #34417
